File: problem/0136_SingleNumber.js
// 136_SingleNumber
// https://leetcode.com/problems/single-number/
// Given a non-empty array of integers nums, every element appears twice except for one. Find that single one.

// You must implement a solution with a linear runtime complexity and use only constant extra space.
/**
 * @param {number[]} nums
 * @return {number}
 */

// brute force

// var singleNumber = function(nums) {
//   const hash ={}
//   for(let i in nums){
//     if(hash[nums[i]] === undefined) {
//       hash[nums[i]] = 0
//     }
//     hash[nums[i]] += 1
//   }
//   console.log(hash)
//   for (let j in hash) {
//     if(hash[j] === 1){
//       return j
//     }
//   }
// };
// time O(n)
// space O(n)

// *************************
// XOR

var singleNumber = function(nums){
  return nums.reduce((p, c) => p^c)
}




console.log(singleNumber([4,1,2,1,2]))

File: c#-solution/0136. Single Number.cs
// 136. Single Number
// https://leetcode.com/problems/single-number/
// Tag: C#, Array, Bit Manipulation, Easy

// bitwise
public class Solution {
    public int SingleNumber(int[] nums) {
        var result = 0;
        for(int i=0; i<nums.Length; i++){
            result ^= nums[i];
        }
        return result;
    }
}
// TC: O(n)
// MC: O(1)

// List
public class Solution
{
  public int SingleNumber(int[] nums)
  {
    List<int> lst = new List<int>();
    for (int i = 0; i < nums.Count(); i++)
    {
      if (lst.Contains(nums[i])) lst.Remove(nums[i]);
      else lst.Add(nums[i]);
    }
    return lst[0];
  }
}

// TC: O(n)
// MC: O(n)